[AMDGPU] Add Early Register Spilling

kmitropoulou opened the pull request:
ERS (Early Register Spilling) tries to spill close to high register pressure. In this way, the spill overhead will impact only the high register pressure blocks. It uses the Register Pressure Tracker to find the parts of the code that have high register pressure. It uses next-use distance analysis to find which register to spill. ERS chooses to spill the live registers with the longest next-use distances (uses that are far away).

We emit the spill instruction as follows:
1. default case: we emit the spill instruction in the high register pressure block
2. emit the spill instruction in the exit block if the high register pressure block is inside a loop
3. emit the spill instruction in the common dominator if there are reachable uses (these are uses that are not dominated by the high register pressure block, but there is a path from high register pressure block to the use)

We optimize the number of the restore instructions as follows:
1. group together uses where the head of the group dominates all the other uses in the group. In this case, we emit 1 restore instruction for the head of the group
2. find an eligible common dominator to emit the restore instruction
3. keep spilled registers live in the low pressure path

ERS is based on ideas from the paper of Matthias Braun and Sebastian Hack: Register Spilling and Live-Range Splitting for
SSA-Form Programs ( https://pp.ipd.kit.edu/publication.php?id=braun09cc )

Commit:
Add Early Register Spilling

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h:
```diff
@@ -0,0 +1,177 @@
+  AMDGPUNextUseAnalysis NUA;
+  /// Keep the registers that are spilled.
+  DenseSet<Register> SpilledRegs;
+  /// keep the output registers of the restored instructions.
```
```suggestion
  /// Keep the output registers of the restored instructions.
```

kmitropoulou replied:
Done

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h:
```diff
@@ -0,0 +1,177 @@
+                    const SetVectorType &UnreachableUses,
+                    const TargetRegisterClass *RC, int FI);
+
+  /// Main spill functions that emits the spill and restore code.
```
```suggestion
  /// Main spill function that emits the spill and restore code.
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h:
```diff
@@ -0,0 +1,177 @@
+
+  /// Emit restore instructions where it is needed
+  MachineInstr *emitRestore(Register SpillReg, MachineInstr *UseMI, int FI);
+  /// Emit restore instruction at the end of a basic block
```
```suggestion
  /// Emit restore instruction at the end of a basic block.
```

kmitropoulou replied:
Done

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h:
```diff
@@ -0,0 +1,177 @@
+    auto It = SpilledRegs.find(Reg);
+    if (It != SpilledRegs.end())
+      return true;
+    return false;
```
```suggestion
    return It != SpilledRegs.end(); 
```

Instead of the whole if (ditto below)

kmitropoulou replied:
Good catch! @jmmartinez did a similar comment and I chose his proposal.

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+  default:
+    return false;
+  }
+  return false;
```
```suggestion
  llvm_unreachable
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+MachineInstr *
+AMDGPUEarlyRegisterSpilling::emitRestore(Register DefRegToSpill,
+                                         MachineBasicBlock *InsertBB, int FI) {
+  assert(InsertBB && "The value of the block is nullptr.");
```
Make it a reference?

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+         !isSpillSaveInstr(CandidateMI) && !isSpillRestoreInstr(CandidateMI) &&
+         !isSpilledReg(CandidateReg) && !isRestoredReg(CandidateReg) &&
+         !CandidateReg.isPhysical() && !TRI->isAGPR(*MRI, CandidateReg) &&
+         !isVCCRegister(CandidateReg) && !CandidateMI->isTerminator() &&
```
Isn't VCC already covered by the check for physical registers?

kmitropoulou replied:
Good catch!

rovka reviewed (commented):
(I'm still making my way through this, but I'm off for lunch and I don't particularly trust github to preserve these comments).

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h:
```diff
@@ -0,0 +1,177 @@
+    auto It = SpilledRegs.find(Reg);
+    if (It != SpilledRegs.end())
+      return true;
+    return false;
```
```suggestion
    return SpilledRegs.contains(Reg);
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+                 FI);
+  }
+  // Reset the tracker because it has already read the next instruction which
+  // we might have modified by a emitting a spill or restore instruction.
```
```suggestion
  // we might have modified by emitting a spill or restore instruction.
```

Commit:
Address review comments

kmitropoulou commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h:
```diff
@@ -0,0 +1,177 @@
+  DenseSet<Register> SpilledRegs;
+  /// keep the output registers of the restored instructions.
+  DenseSet<Register> RestoredRegs;
+  /// Similar to next-use distance analysis, ee assume an approximate trip count
```
Done

kmitropoulou commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+  case AMDGPU::SI_SPILL_V352_SAVE:
+  case AMDGPU::SI_SPILL_V384_SAVE:
+  case AMDGPU::SI_SPILL_V512_SAVE:
+  case AMDGPU::SI_SPILL_V1024_SAVE:
```
I removed this function. I replaced it with isVGPRSpill() from SIInstrInfo.h

kmitropoulou commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+  return Restore;
+}
+
+// TODO: Enable spill spill and restore instructions.
```
No, I updated the comment. I meant that this function needs more tuning.

kmitropoulou commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,940 @@
+    return Pair1.second > Pair2.second;
+  });
+
+  SmallVector<Register> RegistersToSpill;
```
Thank you!

kmitropoulou commented on llvm/test/CodeGen/AMDGPU/test_ers_basic_case.ll:
```diff
@@ -0,0 +1,112 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 6
```
I renamed the tests. Please keep in mind that the tests have multiple spills and restores.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp:
```diff
@@ -0,0 +1,573 @@
+        if (PathMBB == FromMBB)
+          break;
+        auto It = MBBData.find(PathMBB);
+        PathMBB = It != MBBData.end() ? It->second.BestPred : nullptr;
```
Nitpick:
```suggestion
        PathMBB = MBBData.lookup_or(PathMBB, nullptr);
```

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp:
```diff
@@ -0,0 +1,573 @@
+        PathMBB = It != MBBData.end() ? It->second.BestPred : nullptr;
+      }
+      assert(Path.back() == FromMBB && "Incomplete path!");
+      auto *Pred = MBBData[CurMBB].BestPred;
```
At this point, `CurMBB` should be equal to `ToMBB` right ?
```suggestion
      auto *Pred = MBBData[ToMBB].BestPred;
```

jmmartinez reviewed (commented):
Some more comments (sorry it's a big PR)

kmitropoulou commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp:
```diff
@@ -0,0 +1,573 @@
+            (LoopBB && LoopTo && !LoopTo->contains(LoopBB))) {
+          return BB->size() * LoopWeight * MLI->getLoopDepth(BB);
+        } else
+          return BB->size();
```
The https://github.com/llvm/llvm-project/pull/171520 prints for each register of each test the distances, you can double check.

kmitropoulou replied:
Done.

ruiling commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,861 @@
+// This is due to the fact that some unreachable uses might become reachable if
+// we spill in common dominator.
+std::pair<SetVectorType, SetVectorType>
+AMDGPUEarlyRegisterSpilling::collectReachableAndUnreachableUses(
```
I think this is misuse of `reachable`. More accurate description would be `ReachableUsesExcludingDominated` or `NonDominatedReachableUses`. The dominated uses are reachable too.

Commit:
Update next-use distance for uses inside a loop nest with multiple loops at each
loop level.

Commit:
Address review comments

Commit:
Address review comments

kmitropoulou commented on llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp:
```diff
@@ -0,0 +1,861 @@
+    if (SpillBlock == nullptr)
+      continue;
+    // The next step is to check if there are any uses which are reachable
+    // from the SpillBlock. In this case, we have to emit the spill in the
```
In the following example, the spill and the restore will be executed under different execution masks:

```
             +--------------+
             |     BB1      |
             |    r1 = ...  |
             +--------------+
                  /       |
       +--------------+   |
       |      BB2     |   |
       |    spill r1  |   |
       +--------------+   |
                  \       |
             +--------------+
             |     BB3      |
             |              |
             +--------------+
                     |
             +--------------+
             |     BB4      |
             |r1'=restore r1|
             |   ... = r1'  |
             +--------------+
```

kmitropoulou replied:
I agree we have to tune the algorithm. So, we have to explore more options. Spill and restore instructions are expensive. We should find ways to reduce their overhead. Can we fix the problem by adding a spill and restore inside the high register pressure block?

Commit:
Update the loop weight to be 1000 to the power of the loop depth

Currently, the loop weight is the loop depth multiplied by 1000(which represents
the loop iterations).
Now, the loop weight is 1000 to the power of the loop depth.

Commit:
Remove Path from ShortestPathTable

Commit:
Small refactoring

Commit:
Small refactoring

Commit:
Calculate next-use distance on demand

Commit:
Make sorting deterministic

Commit:
Fix live-though calculation
